---
title: "Strain tracking for 1 month samples"
author: "Chunyu Zhao"
date: "01/24/2018"
output: pdf_document
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(
  tidy=FALSE,
  cache=FALSE,
  echo=FALSE,
  message = FALSE,
  warning = FALSE,
  dpi=100,
  fig.width = 8,
  fig.height = 8,
  fig.align = "center",
  cache.lazy = FALSE,
  dev=c("png", "pdf"),
  warning=FALSE)
```

```{r}
library(dplyr)
library(ggplot2)
library(qiimer)
library(readr)
library(tidyr)
library(magrittr)
library(reshape2)
library(pheatmap)
library(pander)
library(broom)

library(subfunc)
library(grid)
library(stringr)
library(vegan)
library(ape)
library(forcats)
library(lubridate)
library(magrittr)

library(colorRamps)
library(RColorBrewer)

library(ggbeeswarm)

library(nlme)

source("helper.R")

require(Biostrings)
library(Rsamtools)
library(stringr)

```

# Introduction

- For each `SubjectID`, we mapped the corresponding 1mon sample reads back to the `contigs_meconium`. 

- So, the contigs info (taxonomic assignment and 139 SCGs alignment) are the one for the meconium samples.

What's new here is: read in the table `sample_conversion_1month.txt`, and update the count/substitution rate.

## Read in the meconium contigs information

```{r}
single_copy_genes_tables <- read.table("sunbeam_output_20171208/single_copy_genes_tables.txt", sep="\t", header=T, stringsAsFactors = F)
single_copy_genes_tables %<>% select(-one_of(c("source","bin_id")))
```

## Read in sample conversion 
```{r}
sample_mapping <- read.table("sunbeam_output_1mo/sample_conversion_1month_20180110.txt", sep="\t", stringsAsFactors = F) %>%
  dplyr::rename(SampleID = V1, SampleID1mo = V2)
```

## start with E coli samples
```{r}
single_copy_genes_tables %<>% filter(SampleID %in% sample_mapping$SampleID)

library(qiimer)

single_copy_genes_tables %<>% filter(! is.na(subject_id))

adf <- single_copy_genes_tables %>% select(superkingdom:species)

#- cuz some of the genus assignments are missing
a <- simplify_assignments(adf, rank1="phylum", rank2="species")
single_copy_genes_tables$Genus <- a
```

# Read in base entropy and substitution rate from sunbeam_neutrino

The following are some cases:

- 1. Some `contigs_meconium`'s `aligned_SCG` (e.g. E coli): zero reads from 1mon mapper. What should we do about these genes?

- 2. There are also some samples's `contigs_meconium` (e.g. s272.STL.V02_1mo): `1mon_reads` mapped to the same contigs, but not the same region (aka gene)

- After talking with Kyle: what we are really interested is `mismatch rate`.

I really don't have a good idea of getting the observed position in the contig string.(sha?)

```{r}
library(binom)

get_lower_upper <- function(x,n){
  confi_inter <- binom.bayes(n-x, n)
  data.frame(lower = confi_inter["lower"], uppder = confi_inter["upper"])
}
```

```{r mismatch rate start from here 20180115, fig.width=20, eval=FALSE}
per_gene_entropy <- function(sampleid, single_copy_genes_tables, sample_mapping){
  ## all the SCGs alignment for birth meconium sample
  temp <- single_copy_genes_tables %>% filter(SampleID %in% sampleid)
  ## corresponding 1-month SampleID
  sampleid_1mon <- sample_mapping %>% filter(SampleID %in% sampleid) %>% select(SampleID1mo) %>% as.character()
  
  ref_file <- file.path(paste("sunbeam_output_20171208/summary/", sampleid, "_base.counts.ent", sep=""))
  input_file <- file.path(paste("sunbeam_output_1mo/summary/", sampleid_1mon, "_base.counts.ent", sep=""))
  
  if (file.exists(input_file)){
    dat <- read.table(input_file,header=T,sep="\t",na.strings="NA",dec=".",strip.white=T,stringsAsFactors = F)
    
    if (nrow(dat) > 1){
      # okay here: what we really want is the mismatch rate compared to the contigs observed bases
      # so, for the same contig for the same position, we need to get the most observed bases
      #! I am being lazy but not really know how to do it in the pipeline anyway. 
      
      observed_base <- read.table(ref_file,header=T,sep="\t",na.strings="NA",dec=".",strip.white=T,stringsAsFactors = F) %>%
        select(ref, pos, A:T) %>%
        gather(obs_base, obs_reads, A:T) %>% 
        group_by(ref, pos) %>% 
        arrange(desc(obs_reads)) %>% 
        filter(row_number() == 1) %>% 
        ungroup()
      
      # okay now it is time to check for the dimensions for dat and observed_base
      # it is kinda expected that the dimensions are different. after all, there are 1mon_reads mapped to other regions of the contig rather than the previously predicted genes
      
      #1 first we only kept regions of contigs covered by the earlier samples
      dat <- merge(dat, observed_base)
      
      #2 calculate the match rate
      dat %<>% gather(obs_1mon_base, obs_1mon_reads, A:T) %>%
        filter(obs_1mon_base == obs_base) %>%
        mutate(mismatchRate = 1 - obs_1mon_reads/n)
      
      
      #- let's forget about the entropy, and do the mismatch rate, with error bar
      pdf_name <- paste(c(sampleid_1mon, "_mismatch_rate.pdf"), collapse="")
      pdf(pdf_name)
      #- we need to consider multiple-copy genes: so we can't use gene names
      hist_sub_rates <- list()
      mismatch_alignment_1month <- list()
      
      ## LOOP over each row of the original `temp` dataframe
      for (x in 1:dim(temp)[1]){
        #print(x)
        #- loop over each single copy genes
        #- I am bit confused about the 0-based coordiates
        
        #- for the 1month sample, it is possible that the previous contig/gene didn't show up 
        one_gene_from_birth_alignment <- temp[x,] %>%
            select(SampleID, gene_name, start, stop, length, query_id, len, Genus)
        
        isExist <- dat %>% filter(ref %in% one_gene_from_birth_alignment$query_id) %>% nrow() > 1
        
        ## case 1
        if (isExist){
          ## hmm alignment region: start - stop
          istart <- as.numeric(one_gene_from_birth_alignment$start) + 1 
          istop <-  as.numeric(one_gene_from_birth_alignment$stop)
          toplot <- dat %>%
            filter(ref %in% one_gene_from_birth_alignment$query_id) %>% 
            filter(pos >=  istart & pos <= istop ) %>%
            mutate(Genus = one_gene_from_birth_alignment$Genus) %>%
            mutate(gene_name = one_gene_from_birth_alignment$gene_name)
          
          ## 20180112: sample-gene-species-(1)mismatch==1 (2)
          ## ABOUT GAPS: if it is important genes, then should not have gaps <- make sense
          
          mismatch_alignment_1month[[x]] <- toplot %>% 
            group_by(ref) %>%
            summarise(mismacthSum = sum(mismatchRate == 1), alignedPos = n()) %>%
            ungroup() %>%
            merge(one_gene_from_birth_alignment) %>%
            select(SampleID, everything())
          
      
          ## case 2
          if (nrow(toplot) > 1){
            confi <- do.call(rbind, lapply(1:nrow(toplot), function(x) get_lower_upper(toplot$obs_1mon_reads[x], toplot$n[x])))
            toplot <- cbind(toplot, confi)
            
            #- okay it seems like I need to think more about this
            fig <- toplot %>% 
              ggplot(aes(x = pos, y = mismatchRate)) +
              geom_point() +
              geom_linerange(aes(ymin = lower, ymax = upper)) + 
              theme_bw() + 
              ylim(0, 1) +
              ggtitle(paste(c(sampleid_1mon, ":", one_gene_from_birth_alignment$gene_name, ":", one_gene_from_birth_alignment$Genus), collapse="")) +
              theme(plot.title = element_text(hjust = 0.5)) +
              theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
              scale_color_brewer(palette = "Set1")
            
            hist_sub_rates[[x]] <- toplot[,c("ref", "pos","mismatchRate","lower","upper","Genus","gene_name")]
            
            print(fig)
            rm(toplot)
          }
        }
      }
      dev.off()
      
      ## mismatch summary for one sample
      mismatch_alignment_1month <- do.call(rbind,mismatch_alignment_1month)
      ## todo: query_id & gene_name IN temp but NOT IN miamatch_alignment_1month
      mis_file = paste(sampleid_1mon, "_mismatch_aln.rda", collapse = "")
      save(mismatch_alignment_1month, file = mis_file)
      
      
      hist_sub_rates <- do.call(rbind, hist_sub_rates)
      hist_file = paste(sampleid_1mon, "_hist_mismatch_rates.rda", collapse = "")
      save(hist_sub_rates, file = hist_file)
      
      ngenus <- length(unique(hist_sub_rates$Genus))
      hist <- hist_sub_rates %>%
        ggplot(aes(x = mismatchRate)) +
        geom_histogram() +
        scale_y_log10() +
        facet_wrap(~Genus, nrow=ngenus, scales="free_y") +
        labs(x ="Substitution rate", y= "Count") +
        theme_bw() + 
        ggtitle("Histogram of mismatch rate over all genes") +
        theme(plot.title = element_text(hjust = 0.5)) +
        #theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
        theme(strip.text.y = element_text(angle = 0))
      
      pdf_name <- paste(c(sampleid_1mon, "_histogram_mismatch_rates.pdf"), collapse="")
      if (ngenus <= 4){
        pdf(pdf_name)
      } else if (ngenus <= 8) {
        pdf(pdf_name, height=14)
      } else {
        pdf(pdf_name, height=22)
      }
        
      print(hist)
      dev.off()
    }
  }
}

samples_all <- unique(single_copy_genes_tables$SampleID)

for ( i in 1:length(samples_all)){
  print(samples_all[i])
  per_gene_entropy(samples_all[i], single_copy_genes_tables, sample_mapping)
}
```

# Mismatch Rate

## first let's read in the 1 month mismatch result

- one sample: multiple contigs aligned to different locations to the same gene => take proportion and aggregate into mean; however, once we aggregate, we won't be able to trace back the alignment length and etc.

```{r}
## samples of birth time point from the anvio single copy genes stats table
samples_all <- unique(single_copy_genes_tables$SampleID)

read_mismatch_file <- function(mis_file){
  if (file.exists(file.path("ret_1mo",mis_file))){
    load(file.path("ret_1mo",mis_file))
    mismatch_alignment_1month %>%
      mutate(Prop = mismacthSum / alignedPos) %>%
      group_by(Genus, gene_name) %>%
      summarise(avgMismatchProp = mean(Prop)) %>%
      ungroup() 
  }
}

total_mismatch_df <- do.call(rbind,lapply(1:length(samples_all), function(i){
  sampleid <- samples_all[i]
  sampleid_1mon <- sample_mapping %>% filter(SampleID %in% sampleid) %>% select(SampleID1mo) %>% as.character()
  mis_file = paste(sampleid_1mon, "_mismatch_aln.rda", collapse = "")
  read_mismatch_file(mis_file) %>%
    mutate(SampleID = sampleid_1mon)
}))

```

```{r}
read_mismatch_file <- function(mis_file){
  if (file.exists(file.path("ret_birth_rerun",mis_file))){
    load(file.path("ret_birth_rerun",mis_file))
    mismatch_alignment_1month %>%
      mutate(Prop = mismacthSum / alignedPos) %>%
      group_by(Genus, gene_name) %>%
      summarise(avgMismatchProp = mean(Prop)) %>%
      ungroup() 
  }
}

total_mismatch_df_birth <- do.call(rbind,lapply(1:length(samples_all), function(i){
  sampleid <- samples_all[i]
  sampleid_1mon <- sampleid
  mis_file = paste(sampleid_1mon, "_mismatch_aln.rda", collapse = "")
  read_mismatch_file(mis_file) %>%
    mutate(SampleID = sampleid_1mon)
}))

```

## Proteobacteria Escherichia coli

### Birth

```{r}
total_mismatch_df_birth %>%
  filter(Genus %in% "Proteobacteria Escherichia coli") %>%
  ggplot(aes(x = gene_name, y = avgMismatchProp, color=SampleID)) +
  geom_point() + 
  theme_bw() +
  scale_y_continuous(labels=scales:::percent) + 
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  facet_wrap(~SampleID, ncol=3) +
  ggtitle("Proteobacteria Escherichia coli")
```

### 1 month

```{r}
total_mismatch_df %>%
  filter(Genus %in% "Proteobacteria Escherichia coli") %>%
  ggplot(aes(x = gene_name, y = avgMismatchProp, color=SampleID)) +
  geom_point() + 
  theme_bw() +
  scale_y_continuous(labels=scales:::percent) + 
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  facet_wrap(~SampleID, ncol=3) +
  ggtitle("Proteobacteria Escherichia coli")
```

## Bacteroidetes Bacteroides vulgatus

### Birth

```{r}
total_mismatch_df_birth %>%
  filter(Genus %in% "Bacteroidetes Bacteroides vulgatus") %>%
  ggplot(aes(x = gene_name, y = avgMismatchProp, color=SampleID)) +
  geom_point() + 
  theme_bw() +
  scale_y_continuous(labels=scales:::percent) + 
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  facet_wrap(~SampleID, ncol=3) +
  ggtitle("Bacteroidetes Bacteroides vulgatus")
```

### 1 month

```{r}
total_mismatch_df %>%
  filter(Genus %in% "Bacteroidetes Bacteroides vulgatus") %>%
  ggplot(aes(x = gene_name, y = avgMismatchProp, color=SampleID)) +
  geom_point() + 
  theme_bw() +
  scale_y_continuous(labels=scales:::percent) + 
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  facet_wrap(~SampleID, ncol=3) +
  ggtitle("Bacteroidetes Bacteroides vulgatus")
```

## Firmicutes Enterococcus faecalis

### Birth

```{r}
total_mismatch_df_birth %>%
  filter(Genus %in% "Firmicutes Enterococcus faecalis") %>%
  ggplot(aes(x = gene_name, y = avgMismatchProp, color=SampleID)) +
  geom_point() + 
  theme_bw() +
  scale_y_continuous(labels=scales:::percent) + 
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  facet_wrap(~SampleID, ncol=3) +
  ggtitle("Firmicutes Enterococcus faecalis")
```

### 1 month

```{r}
total_mismatch_df %>%
  filter(Genus %in% "Firmicutes Enterococcus faecalis") %>%
  ggplot(aes(x = gene_name, y = avgMismatchProp, color=SampleID)) +
  geom_point() + 
  theme_bw() +
  scale_y_continuous(labels=scales:::percent) + 
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  facet_wrap(~SampleID, ncol=3) +
  ggtitle("Firmicutes Enterococcus faecalis")
```
