---
title: "Tree + Panphlan"
author: "Chunyu Zhao"
date: \today
output: pdf_document
editor_options: 
  chunk_output_type: console
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(
  tidy=FALSE,
  cache=FALSE,
  echo=FALSE,
  message = FALSE,
  warning = FALSE,
  dpi=100,
  fig.width = 8,
  fig.height = 8,
  fig.align = "center",
  cache.lazy = FALSE,
  dev=c("png", "pdf"),
  warning=FALSE)
```

```{r}
library(tidyverse)
library(pander)
library(ggplot2)
library(ggtree)
library(ape)
library(scales)
library(colorRamps)
library(RColorBrewer)
library(magrittr)
library(ggbeeswarm)
library(readr)
library(viridis)
library(ggsci)
library(qiimer)
library(ape)
library(vegan)
library(ggplot2)
library(dplyr)
library(reshape2)
library(kylemisc)
library(tidyr)
library(broom)
library(pheatmap)

library(subfunc)
library(forcats)
library(lubridate)
library(magrittr)
library(pander)
library(ggbeeswarm)

library(RColorBrewer)
library(readr)
```

```{r}
source("helper.R")

read_pan_matrix <- function(filepath, commented = TRUE, metadata = TRUE) {
  f <- file(filepath, "rt")
  header_line <- readLines(f, n = 1)
  
  if (commented){
    header_line <- readLines(f, n = 1)
  }
  
  col_names <- strsplit(header_line, "\t")[[1]]
  col_classes <- rep("numeric", times = length(col_names))
  col_classes[1] <- "character"
  
  full_otu_table <- read.table(f, col.names = col_names, colClasses = col_classes,
                               sep = "\t", quote = "", as.is = TRUE, header = FALSE)
  close(f)
  
  data_cols <- 2:length(col_names)
  
  sample_ids <- col_names[data_cols]
  otu_ids <- as.character(full_otu_table[, 1])
  counts <- as.matrix(full_otu_table[, data_cols])
  rownames(counts) <- otu_ids
  
  list(sample_ids = sample_ids, otu_ids = otu_ids, counts = counts)
}
```

# E. faecalis

## Phylogenetic Tree

- We read the `external genomes` list from the **anvio** pangenome analysis results.
- Read in the tree from **panphlan** 
- write to `20180720_efaecalis_tree_tip_order.txt`.

```{r}
## read in genome list from anvio
external_genomes <- 
  read_delim(file.path("efaecalis/", "external-genomes-all.txt"), delim="\t") %>%
  separate(contigs_db_path, into="accession", extra="drop", sep="\\.", remove=FALSE) %>%
  dplyr::rename(taxon=name)

dd <- external_genomes %>%
  mutate(isPCMP = ifelse(grepl("^s", taxon), TRUE, FALSE)) %>%
  mutate(node_lable = gsub("E_faecalis_strain_","",taxon))
row.names(dd) <- NULL

dd %>% write.table(file.path("efaecalis/", "20180720_efaecalis_tree_dd.txt"),sep="\t", quote=FALSE, row.names = F)

tree <- read.tree(file.path("efaecalis", "tree.txt"))

## re-label roder the tree tip lables
tree2 <- ladderize(tree, right = FALSE)
#tree$tip.label
#tree2$tip.label
plot(tree2)
nodelabels()
tiplabels()

is_tip <- tree2$edge[,2] <= length(tree2$tip.label)
ordered_tips <- tree2$edge[is_tip, 2]

## read in tree order we want
filename <- file.path("efaecalis/", "20180720_efaecalis_tree_tip_order.txt")
write.table(tree2$tip.label[ordered_tips], file=filename, sep="\t", quote=F, row.names = F, col.names = F)
```

### Tree time

- generate `20180702_efaecalis_tree_v1.pdf` => KB18 supple files

```{r}
mycolor <- pal_npg(palette = c("nrc"), alpha = 1)(10)
treefile <- file.path("efaecalis/", "20180702_efaecalis_tree_v1.pdf")

p <- ggtree(tree)

p %<+% dd +
  geom_tippoint(aes(subset = isPCMP, shape = isPCMP, color = node_lable,  x = x + 0.0001), size=2) + 
  geom_tiplab(aes(subset = !isPCMP, label = node_lable), size=2.5) +
  scale_shape_manual(values = c( 15, 16), guide = FALSE) +
  theme(legend.position="right") + 
  scale_color_viridis(option = "magma", discrete = TRUE) +
  ggsave(treefile, width=5, height=4.5, useDingbats=FALSE)
```

## Panphlan

- visualize pangenome results
- read in `result_gene_presence_absence.csv`
- read in `20180720_efaecalis_tree_tip_order.txt` 

```{r}
o <- read_pan_matrix("efaecalis/result_gene_presence_absence.csv", commented=FALSE)
cts <- o$counts
props <- sweep(cts, 2, colSums(cts), "/")

strain_names <- colnames(cts)
gene_names <- rownames(cts)

tree.tip.order <- read_delim("efaecalis/20180720_efaecalis_tree_tip_order.txt", delim="\t", col_names = F) 
tree.tip.order <- tree.tip.order$X1
```

### Heatmap 

- 20180720_Fig2A_efaecalis_pan_heatmap.pdf

```{r}
cts.toplot <- cts
colnames(cts.toplot) <- sub("REF_GCF","GCF",colnames(cts.toplot))
cts.toplot <- cts.toplot[, dd$accession]

if (length(setdiff(colnames(cts.toplot), dd$accession)) > 1)
  print("WARNING: labels don't match")

dd.heatmap <- dd %>% 
  select(accession, taxon, node_lable) %>%
  mutate(taxon = factor(taxon, levels= tree.tip.order)) %>% 
  filter(accession %in% colnames(cts.toplot)) %>% 
  mutate(taxon = factor(taxon, levels = tree.tip.order)) %>% 
  arrange(taxon) %>%
  mutate(accession = factor(accession, levels = .$accession))

color = colorRampPalette(brewer.pal(n = 9, name = "RdPu"))(9)
color <- c("white", color[c(4)])
breaks = seq(0, 1, length.out = 3)

# clustering
row.order <- hclust(dist(cts.toplot))$order 
col.order <- levels(dd.heatmap$accession)
# re-order matrix accoring to clustering
dat_new <- cts.toplot[row.order, col.order] 

# melt to dataframe
df_molten_dat <- reshape2::melt(as.matrix(dat_new)) # reshape into dataframe
names(df_molten_dat)[c(1:3)] <- c("gene_family", "genome", "counts")

df_molten_dat %>% 
  mutate(counts = as.factor(counts)) %>%
  ggplot(aes(x = gene_family, y = genome, fill = counts)) + 
  geom_tile() +
  scale_fill_manual(values = color) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  theme_minimal() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks=element_blank()) +
  ggsave("efaecalis/20180720_Fig2A_efaecalis_pan_heatmap.pdf", width=18, height=8, useDingbats=FALSE)
```

### PCoA plot based on Jacard Distance

- 20180720_efaecalis_pcoa.pdf

```{r}
props.toplot <- props
colnames(props.toplot) <- sub("REF_GCF","GCF",colnames(props.toplot))
props.toplot <- props.toplot[, colnames(props.toplot) %in% as.character(dd$accession)]
## good: accession being factor causing some troubles

jc <- dist(t(props.toplot), method = "binary")
pc <- pcoa(jc)

matrix_to_sample_df <- function (x) {
  df <- data.frame(SampleID=rownames(x), stringsAsFactors = FALSE)
  cbind(df, x)
}

pc_df <- matrix_to_sample_df(pc$vectors[,1:3])
pc_df %<>% merge(dd, by.x="SampleID", by.y="accession")

#setdiff(dd$accession %>% as.character(), pc_df$SampleID)

pc_pct <- round(pc$values$Relative_eig * 100)

mycolor <- pal_npg(palette = c("nrc"), alpha = 1)(10)

pc_df %<>% 
  mutate(isPCMP = ifelse(grepl("^s", taxon), TRUE, FALSE)) %>%
  mutate(isPCMP = as.factor(isPCMP))

fig <- pc_df %>% 
  ggplot(aes(x=Axis.1, y=Axis.2, color=isPCMP, shape=isPCMP)) +
  geom_point(aes_string(x="Axis.1", y="Axis.2")) +
  theme_bw() +
  labs(x = paste0("PCoA axis 1 (", pc_pct[1], "%)"), y = paste0("PCoA axis 2 (", pc_pct[2], "%)")) +
  coord_equal() +
  theme(plot.title = element_text(hjust = 0.5)) +
  scale_color_npg() +
  scale_shape_manual(values = c(16, 8), guide=FALSE) +
  geom_text(aes(label=ifelse(isPCMP == TRUE, as.character(SampleID),'')),hjust=0,vjust=0) 

fig + ggsave("efaecalis/20180720_efaecalis_pcoa.pdf", width=5, height=4.5, useDingbats=FALSE)
```

### Fisher's exact test

- 20180720_efaecalis_fisher.pdf

```{r, fig.width=5, fig.height=5}
#https://github.com/road2stat/ggsci
props_toTest <- props.toplot

genus_present <- apply(props_toTest, 2, function (x) x > 0) %>%
  melt(c("Gene", "SampleID"), value.name="Present") %>%
  mutate(SampleID = sub("REF_", "", SampleID)) %>% 
  merge(dd, by.x="SampleID", by.y = "accession") %>%
  mutate(Group = ifelse(isPCMP, "PCMP", "Non-PCMP")) %>% 
  mutate(Group = factor(Group, levels=c("PCMP", "Non-PCMP")))

low_lm <- 2
high_lm <- ncol(props_toTest)-2
genus_present %<>%
  group_by(Gene) %>%
  filter(sum(Present) < high_lm & sum(Present) > low_lm) %>%
  ungroup()

ret <- genus_present %>%
  group_by(Gene) %>%
  filter(sum(Present) > low_lm, sum(Present) < high_lm) %>%
  do(tidy(fisher.test(with(., table(Present, Group))))) %>%
  ungroup() %>%
  arrange(p.value) %>%
  mutate(fdr_corrected = p.adjust(p.value, method="fdr")) %>%
  filter(p.value <= 0.05) 

ret.sig <- ret %>% 
  arrange(fdr_corrected) %>%
  mutate(isSig = ifelse(fdr_corrected <= 0.05, "*", "")) %>%
  slice(1:10)

pander(ret.sig %>% select(-one_of("method", "alternative")))

fig <- genus_present %>% 
  inner_join(ret.sig, "Gene") %>%
  mutate(Gene = paste(Gene, isSig, sep="")) %>% 
  mutate(Gene = as.factor(Gene)) %>%
  group_by(Gene, Group) %>%
  mutate(frac_present = sum(Present) / n()) %>% 
  ungroup() %>% 
  mutate(Gene = reorder(droplevels(Gene), -p.value)) %>%
  ggplot() +
  geom_bar(
    aes(x=Gene, y=frac_present, fill=Group), 
    position="dodge", stat="identity") +
  scale_fill_brewer(palette="Paired", guide=guide_legend(reverse=T)) +
  theme_bw() +
  labs(y="Fraction of samples where taxon is present", fill="", x="") +
  coord_flip()

fig + ggsave("efaecalis/20180720_efaecalis_fisher.pdf", width=7, height=5, useDingbats=FALSE)
```

# B. vulgatus

## Phylogenetic Tree

```{r}
external_genomes <- read_delim(file.path("bvulgatus/", "external-genomes-all.txt"), delim="\t") %>%
  separate(contigs_db_path, into="accession", extra="drop", sep="\\.", remove=FALSE) %>%
  dplyr::rename(taxon=name)

dd <- external_genomes %>%
  mutate(isPCMP = ifelse(grepl("^s", taxon), TRUE, FALSE)) %>%
  mutate(node_lable = gsub("E_faecalis_strain_","",taxon))
row.names(dd) <- NULL

dd %>% write.table(file.path("bvulgatus/", "20180720_bvulgatus_tree_dd.txt"),sep="\t", quote=FALSE, row.names = F)

tree <- read.tree(file.path("bvulgatus", "tree.txt"))

tree2 <- ladderize(tree, right = FALSE)
tree$tip.label
tree2$tip.label
plot(tree2)
nodelabels()
tiplabels()

is_tip <- tree2$edge[,2] <= length(tree2$tip.label)
ordered_tips <- tree2$edge[is_tip, 2]

filename <- file.path("bvulgatus/", "20180720_bvulgatus_tree_tip_order.txt")
write.table(tree2$tip.label[ordered_tips], file=filename, sep="\t", quote=F, row.names = F, col.names = F)
```

### tree time

- 20180702_bvulgatus_tree_v1.pdf

```{r}
mycolor <- pal_npg(palette = c("nrc"), alpha = 1)(10)
treefile <- file.path("bvulgatus/", "20180702_bvulgatus_tree_v1.pdf")

p <- ggtree(tree)

p %<+% dd +
  geom_tippoint(aes(subset = isPCMP, shape = isPCMP, color = node_lable,  x = x + 0.0001), size=2) + 
  geom_tiplab(aes(subset = !isPCMP, label = node_lable), size=2.5) +
  scale_shape_manual(values = c( 15, 16), guide = FALSE) +
  theme(legend.position="right") + 
  scale_color_viridis(option = "magma", discrete = TRUE) +
  ggsave(treefile, width=5, height=4.5, useDingbats=FALSE)
```

## Panphlan

```{r}
o <- read_pan_matrix("bvulgatus/result_gene_presence_absence.csv", commented=FALSE)
cts <- o$counts
props <- sweep(cts, 2, colSums(cts), "/")

strain_names <- colnames(cts)
gene_names <- rownames(cts)

tree.tip.order <- read_delim("bvulgatus/20180720_bvulgatus_tree_tip_order.txt", delim="\t", col_names = F) 
tree.tip.order <- tree.tip.order$X1
```

### Heatmap 

- 20180720_Fig2A_bvulgatus_pan_heatmap.pdf

```{r}
cts.toplot <- cts
colnames(cts.toplot) <- sub("REF_GCF","GCF",colnames(cts.toplot))
cts.toplot <- cts.toplot[, dd$accession]

dd.heatmap <- dd %>% 
  select(accession, taxon, node_lable) %>%
  mutate(taxon = factor(taxon, levels= tree.tip.order)) %>% 
  filter(accession %in% colnames(cts.toplot)) %>% 
  mutate(taxon = factor(taxon, levels = tree.tip.order)) %>% 
  arrange(taxon) %>%
  mutate(accession = factor(accession, levels = .$accession))

color = colorRampPalette(brewer.pal(n = 9, name = "RdPu"))(9)
color <- c("white", color[c(4)])
breaks = seq(0, 1, length.out = 3)

# re-order matrix accoring to clustering
row.order <- hclust(dist(cts.toplot))$order
col.order <- levels(dd.heatmap$accession)
dat_new <- cts.toplot[row.order, col.order]

# melt to dataframe
df_molten_dat <- reshape2::melt(as.matrix(dat_new)) # reshape into dataframe
names(df_molten_dat)[c(1:3)] <- c("gene_family", "genome", "counts")

df_molten_dat %>% 
  mutate(counts = as.factor(counts)) %>%
  ggplot(aes(x = gene_family, y = genome, fill = counts)) + 
  geom_tile() +
  scale_fill_manual(values = color) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  theme_minimal() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks=element_blank()) +
  ggsave("bvulgatus/20180720_Fig2A_bvulgatus_pan_heatmap.pdf", width=18, height=8, useDingbats=FALSE)
```

### PCoA plot based on Jacard Distance

- 20180720_bvulgatus_pcoa.pdf

```{r}
props.toplot <- props
colnames(props.toplot) <- sub("REF_GCF","GCF",colnames(props.toplot))
props.toplot <- props.toplot[, colnames(props.toplot) %in% as.character(dd$accession)]

jc <- dist(t(props.toplot), method = "binary")
pc <- pcoa(jc)

matrix_to_sample_df <- function (x) {
  df <- data.frame(SampleID=rownames(x), stringsAsFactors = FALSE)
  cbind(df, x)
}

pc_df <- matrix_to_sample_df(pc$vectors[,1:3])

pc_df %<>% merge(dd, by.x="SampleID", by.y="accession")

setdiff(dd$accession %>% as.character(), pc_df$SampleID)

pc_pct <- round(pc$values$Relative_eig * 100)

mycolor <- pal_npg(palette = c("nrc"), alpha = 1)(10)

pc_df %<>% 
  mutate(isPCMP = ifelse(grepl("^s", taxon), TRUE, FALSE)) %>%
  mutate(isPCMP = as.factor(isPCMP))

fig <- pc_df %>% 
  ggplot(aes(x=Axis.1, y=Axis.2, color=isPCMP, shape=isPCMP)) +
  geom_point(aes_string(x="Axis.1", y="Axis.2")) +
  theme_bw() +
  labs(x = paste0("PCoA axis 1 (", pc_pct[1], "%)"), y = paste0("PCoA axis 2 (", pc_pct[2], "%)")) +
  coord_equal() +
  theme(plot.title = element_text(hjust = 0.5)) +
  scale_color_npg() +
  scale_shape_manual(values = c(16, 8), guide=FALSE) +
  geom_text(aes(label=ifelse(isPCMP == TRUE, as.character(SampleID),'')),hjust=0,vjust=0) 

fig + ggsave("bvulgatus/20180720_bvulgatus_pcoa.pdf", width=5, height=4.5, useDingbats=FALSE)
```

### Fisher's exact test

- 20180720_bvulgatus_fisher.pdf

```{r, fig.width=5, fig.height=5}
props_toTest <- props.toplot

genus_present <- apply(props_toTest, 2, function (x) x > 0) %>%
  melt(c("Gene", "SampleID"), value.name="Present") %>%
  mutate(SampleID = sub("REF_", "", SampleID)) %>% 
  merge(dd, by.x="SampleID", by.y = "accession") %>%
  mutate(Group = ifelse(isPCMP, "PCMP", "Non-PCMP")) %>% 
  mutate(Group = factor(Group, levels=c("PCMP", "Non-PCMP")))

low_lm <- 2
high_lm <- ncol(props_toTest)-2
genus_present %<>%
  group_by(Gene) %>%
  filter(sum(Present) < high_lm & sum(Present) > low_lm) %>%
  ungroup()

ret <- genus_present %>%
  group_by(Gene) %>%
  filter(sum(Present) > low_lm, sum(Present) < high_lm) %>%
  do(tidy(fisher.test(with(., table(Present, Group))))) %>%
  ungroup() %>%
  arrange(p.value) %>%
  mutate(fdr_corrected = p.adjust(p.value, method="fdr")) %>%
  filter(p.value <= 0.05) 

ret.sig <- ret %>% 
  arrange(fdr_corrected) %>%
  mutate(isSig = ifelse(fdr_corrected <= 0.05, "*", "")) %>%
  slice(1:10)

pander(ret.sig %>% select(-one_of("method", "alternative")))

fig <- genus_present %>% 
  inner_join(ret.sig, "Gene") %>%
  mutate(Gene = paste(Gene, isSig, sep="")) %>% 
  mutate(Gene = as.factor(Gene)) %>%
  group_by(Gene, Group) %>%
  mutate(frac_present = sum(Present) / n()) %>% 
  ungroup() %>% 
  mutate(Gene = reorder(droplevels(Gene), -p.value)) %>%
  ggplot() +
  geom_bar(
    aes(x=Gene, y=frac_present, fill=Group), 
    position="dodge", stat="identity") +
  scale_fill_brewer(palette="Paired", guide=guide_legend(reverse=T)) +
  theme_bw() +
  labs(y="Fraction of samples where taxon is present", fill="", x="") +
  coord_flip()

fig + ggsave("bvulgatus/20180720_bvulgatus_fisher.pdf", width=7, height=5, useDingbats=FALSE)
```
